### ISSUE TITLE
False Negative SyntaxError with unparenthesized assignment expression in dictionary expression/comprehension

### ISSUE THREAD

**Issue by BeefaloKing** [microsoft/pyright#7694]
**Describe the bug**
Use of an unparenthesized assignment expression when defining a dictionary (or dictionary comprehension) results in no problems detected. However, python reports it as a syntax error.

**Code or Screenshots**
python version
```
Python 3.12.2 (main, Feb 21 2024, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)] on linux
```

test.py
```python
foo = {tmp := 'foobar': tmp}
```
```
$ python3 test.py
  File "<...>/test.py", line 1
    foo = {tmp := 'foobar': tmp}
                          ^
SyntaxError: invalid syntax
```

test2.py
```python
strs = ('a', 'b', 'c')
bar = {tmp := s: tmp for s in strs}
```
```
$ python3 test2.py
  File "<...>/test2.py", line 2
    bar = {tmp := s: tmp for s in strs}
                   ^
SyntaxError: invalid syntax
```

Both of these are valid syntax if the assignment expression is surrounded in parentheses.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

Happens in the pyright playground against latest (1.1.358) against python version 3.12
https://pyright-play.net/?pythonVersion=3.12&strict=true&code=MQAgLgpgzmB0AOBPAsAKAGYHtMgLwgG8wBbeEALnwHItMAjAQwCcrzxSBfNNUSGAJgQpUMJlDwgAFFQZUANCCp15igMZUAlGkZMJRUhXxQ2JMll3iAlgDsQoqByA

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.359, which I just published.
